Source/Core/Core/AchievementManager.cpp

```diff
@@ -152,66 +158,218 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
           },
       .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
   rc_hash_init_custom_filereader(&volume_reader);
-  if (!rc_hash_generate_from_file(m_game_hash.data(), RC_CONSOLE_GAMECUBE, iso_path.c_str()))
+  m_queue.EmplaceItem([this, callback, file_path] {
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, file_path.c_str()))
+    {
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from game file {}.",
+                    file_path);
+      callback(AchievementManager::ResponseType::MALFORMED_OBJECT);
+    }
+    {
+      std::lock_guard lg{m_lock};
+      m_game_hash = std::move(new_hash);
+    }
+    LoadGameSync(callback);
+  });
+}
+
+void AchievementManager::HashGame(const DiscIO::Volume* volume, const ResponseCallback& callback)
+{
+  if (!m_is_runtime_initialized)
   {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from game file.");
+    ERROR_LOG_FMT(ACHIEVEMENTS,
+                  "Attempted to load game achievements without Achievement Manager initialized.");
+    callback(AchievementManager::ResponseType::MANAGER_NOT_INITIALIZED);
     return;
   }
-  m_queue.EmplaceItem([this, callback] {
-    const auto resolve_hash_response = ResolveHash(this->m_game_hash);
-    if (resolve_hash_response != ResponseType::SUCCESS || m_game_id == 0)
+  if (volume == nullptr)
+  {
+    INFO_LOG_FMT(ACHIEVEMENTS, "New volume is empty.");
+    return;
+  }
+  if (m_disabled)
+  {
+    INFO_LOG_FMT(ACHIEVEMENTS, "Achievement Manager is disabled until core is rebooted.");
+    OSD::AddMessage("Achievements are disabled until core is rebooted.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    return;
+  }
+  {
+    std::lock_guard lg{m_lock};
+    if (m_loading_volume.get() != nullptr)
     {
-      callback(resolve_hash_response);
-      INFO_LOG_FMT(ACHIEVEMENTS, "No RetroAchievements data found for this game.");
-      OSD::AddMessage("No RetroAchievements data found for this game.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
+      INFO_LOG_FMT(ACHIEVEMENTS, "Disabling Achievement Manager due to hash spam.");
+      DisableRestoreAchievementManager(true);
+      callback(AchievementManager::ResponseType::EXPIRED_CONTEXT);
       return;
     }
-
-    const auto start_session_response = StartRASession();
-    if (start_session_response != ResponseType::SUCCESS)
+    m_loading_volume = DiscIO::CreateVolume(volume->GetBlobReader().CopyReader());
```

Comment:
This entire construct makes me sad but I don't have any better ideas either with the currently available functions in rcheevos.